keps/sig-api-machinery/5073-declarative-validation-with-validation-gen/README.md

```diff
@@ -948,6 +988,190 @@ The linter, as previously described, will enforce rules to address valid zero-va
+   <td>Optional field that once set must remain</td>
+ </tr>
+ <tr>
+   <td>immutable</td>
+   <td>Can be unset</td>
+   <td>set</td>
+   <td>modify, clear</td>
+   <td>Can be set once (now or later), then frozen</td>
+   <td>Field can be set once by user/controller</td>
+ </tr>
+ <tr>
+   <td>required+immutable</td>
+   <td>Must be present</td>
+   <td>none</td>
+   <td>set, modify, clear</td>
+   <td>Must be set at parent creation, value frozen forever</td>
+   <td>Identity field set at creation</td>
+ </tr>
+ <tr>
+   <td>frozen</td>
+   <td>Can be unset</td>
+   <td>none</td>
+   <td>set, modify, clear</td>
+   <td>Field state (set/unset) and value frozen at parent creation</td>
+   <td>Architectural decision at creation</td>
+ </tr>
+</table>
+
+#### Immutability and Lifecycle Patterns Demonstrated
+
+```go
+type DeploymentSpec struct {
+    // +k8s:required // Must exist, can be modified
+    // ✅ Create: replicas = 3
+    // ✅ Update: replicas = 5
+    // ❌ Update: replicas = nil
+    Replicas *int32
+}
+
+type PersistentVolumeStatus struct {
+    // +k8s:requiredOnceSet
+    // ✅ Create: phase = ""              (volume created)
+    // ✅ Update: phase = "Available"     (set once)
+    // ✅ Update: phase = "Bound"         (modify)
+    // ❌ Update: phase = ""              (cannot clear)
+    Phase PersistentVolumePhase
+}
+
+type PersistentVolumeClaimSpec struct {
+    // +k8s:immutable
+    // ✅ Create: volumeName = ""
+    // ✅ Update: volumeName = "pv-123" (set once by PV controller)
+    // ❌ Update: volumeName = "pv-456" (cannot change binding)
+    // ❌ Update: volumeName = "" (cannot unbind)
+    VolumeName string
+}
+
+type PersistentVolumeSpec struct {
+    // required+immutable: Must specify capacity at creation
+    // +k8s:required
+    // +k8s:immutable
+    // ✅ Create: capacity = {"storage": "10Gi"}
+    // ❌ Update: capacity = {"storage": "20Gi"}
+    Capacity ResourceList
+}
+
+type PodSpec struct {
+    // +k8s:frozen // State decided at creation (can be set or unset)
+    // Create with: hostNetwork = true   -> Frozen true
+    // Create with: hostNetwork = false  -> Frozen false
+    // Create without setting            -> Frozen false (zero value)
+    HostNetwork bool
+}
+```
+
+#### List Operations
+
+Lists have two levels of constraints:
+
+**List-Level Constraints:**
+```go
+type ExampleStatus struct {
+    // Array must exist (can be empty)
+    // +k8s:required
+    Conditions []Condition `json:"conditions"`
+}
+```
+
+**Item-Level Constraints:**
+For list items, clearing (nil-ing) a list is represented as unsetting all items. Item-level constraints can prevent list becoming nil:
+
+```go
+type CertificateSigningRequestStatus struct {
+    // List is optional, but certain items are permanent once added
+    // +k8s:optional
+    // +k8s:listType=map
+    // +k8s:listMapKey=type
+    // +k8s:listMapItem([["type","Approved"]])=+k8s:immutable
+    // +k8s:listMapItem([["type","Denied"]])=+k8s:immutable
```

Comment:
confirming my understanding. This means that the item cannot be removed, but the content of the item (message for instance) can change.